### PR TITLE
Allow `source` to be a local path within a run's artifact directory in create_model_version

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Allow `source` to be a local path within a run's artifact directory in create_model_version.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/_model_registry/fluent.py [hunk 1]
```diff
@@ -77,10 +77,9 @@ def register_model(
             raise e
 
     if RunsArtifactRepository.is_runs_uri(model_uri):
-        source = RunsArtifactRepository.get_underlying_uri(model_uri)
         (run_id, _) = RunsArtifactRepository.parse_runs_uri(model_uri)
         create_version_response = client.create_model_version(
-            name, source, run_id, tags=tags, await_creation_for=await_registration_for
+            name, model_uri, run_id, tags=tags, await_creation_for=await_registration_for
```
Comment (dbczumar) [inline mlflow/tracking/_model_registry/fluent.py:82]: Can we check on Databricks to make sure this doesn't break anything? I think it might break model version creation on Databricks.

Comment (dbczumar) [reply mlflow/tracking/_model_registry/fluent.py:82]: I think Databricks expects the URI to be `dbfs:/databricks/mlflow-tracking`....

Comment (dbczumar) [reply mlflow/tracking/_model_registry/fluent.py:82]: Maybe we should instead update the validation logic in MLflow server to check if the source location is within the run's artifact directory.

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -1363,6 +1363,23 @@ def _delete_registered_model_tag():
...
 
 
+def _validate_source(source: str, run_id: str) -> None:
+    if not is_local_uri(source):
+        return
+
+    if run_id:
+        store = _get_tracking_store()
+        run = store.get_run(run_id)
+        if source.startswith(run.info.artifact_uri):
```
Comment (harupy) [inline mlflow/server/handlers.py:1373]: This line needs to be fixed. `/a/b/c/d` (source) should be considered within `file:////a/b/c` (run.info.artifact_uri)

**Comment by harupy** [comment]
TODO: cherry-pick this into branch-2.2

File: mlflow/server/handlers.py [hunk 3]
```diff
@@ -1363,6 +1364,25 @@ def _delete_registered_model_tag():
...
+    if not is_local_uri(source):
+        return
+
+    if run_id:
+        store = _get_tracking_store()
+        run = store.get_run(run_id)
+        source = pathlib.Path(local_file_uri_to_path(source))
+        run_artifact_dir = pathlib.Path(local_file_uri_to_path(run.info.artifact_uri))
+        if run_artifact_dir in [source, *source.parents]:
+            return
```
Comment (WeichenXu123) [inline mlflow/server/handlers.py:1377]: Can we add some unit tests for this part ?
for both UNIX and Windows.

Comment (harupy) [reply mlflow/server/handlers.py:1377]: tests/tracking/test_rest_tracking.py runs both on Unix and Windows.

**Review by WeichenXu123 — APPROVED** [review]

File: mlflow/server/handlers.py [hunk 4]
```diff
@@ -1363,6 +1364,25 @@ def _delete_registered_model_tag():
...
+
+    raise MlflowException(
+        f"Invalid source: '{source}'. source must be a local path within "
+        "a run's artifact directory or a non-local path.",
+        INVALID_PARAMETER_VALUE,
+    )
```
Comment (dbczumar) [inline mlflow/server/handlers.py:1383]: Can we update the message to make it obvious that, if the source is a local path, it has to be contained within the artifact directory of the run specified by the `run_id` request parameter?

File: tests/tracking/test_rest_tracking.py [hunk 5]
```diff
@@ -1047,16 +1047,58 @@ def get(self, key, default=None):
 def test_create_model_version_with_local_source(mlflow_client):
```
Comment (dbczumar) [inline tests/tracking/test_rest_tracking.py:1047]: @harupy Can we also add a test for the case where a user starts the MLflow server with `--default-artifact-root <local_path>` and then a client does `mlflow.sklearn.log_model(..., registered_model_name="mymodel")`? That way, we can catch this kind of regression in the future. Right now, we're missing coverage.

File: mlflow/server/handlers.py [hunk 6]
```diff
@@ -1363,6 +1364,25 @@ def _delete_registered_model_tag():
...
+def _validate_source(source: str, run_id: str) -> None:
+    if not is_local_uri(source):
+        return
+
+    if run_id:
+        store = _get_tracking_store()
+        run = store.get_run(run_id)
+        source = pathlib.Path(local_file_uri_to_path(source))
+        run_artifact_dir = pathlib.Path(local_file_uri_to_path(run.info.artifact_uri))
+        if run_artifact_dir in [source, *source.parents]:
```
Comment (dbczumar) [inline mlflow/server/handlers.py:1376]: If `run_artifact_dir` isn't normalized, this seems like it might not work. For example:

```
In [1]: import pathlib

In [2]: p = pathlib.Path.cwd()

In [3]: p2 = pathlib.PosixPath("/private/tmp/dir1")

In [4]: p
Out[4]: PosixPath('/private/tmp/dir1/dir2/dir3')

In [5]: p2 in [*p.parents]
Out[5]: True

In [6]: p3 = pathlib.PosixPath("/private/tmp/dir1/dir2/..")

In [7]: p3 in [*p.parents]
Out[7]: False
```

File: tests/tracking/test_rest_tracking.py (hunk 5 above)
Comment (harupy) [reply tests/tracking/test_rest_tracking.py:1047]: Added!

File: mlflow/server/handlers.py (hunk 6 above)
Comment (harupy) [reply mlflow/server/handlers.py:1376]: fixed!

File: mlflow/server/handlers.py (hunk 4 above)
Comment (harupy) [reply mlflow/server/handlers.py]: updated!
